### PR TITLE
gh-111968: Use per-thread slice_cache in free-threading

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-111968

### CODE REVIEW

File: Include/internal/pycore_freelist.h [hunk 1]
```diff
@@ -59,10 +59,19 @@ struct _Py_float_state {
...
+    PySliceObject *slice_cache;
+#endif
+};
+
 typedef struct _Py_freelist_state {
     struct _Py_float_state float_state;
     struct _Py_tuple_state tuple_state;
     struct _Py_list_state list_state;
+    struct _Py_slice_state slice_state;
 } _PyFreeListState;
```
Comment (corona10) [inline Include/internal/pycore_freelist.h:75]: @erlend-aasland @colesbury 
IMO, we should rename _PyFreeListState if we consider slice_cache is not freelist :)
Do you have any good suggestion?

File: Objects/sliceobject.c [hunk 2]
```diff
@@ -103,12 +103,20 @@ PyObject _Py_EllipsisObject = _PyObject_HEAD_INIT(&PyEllipsis_Type);
...
+{
+    PySliceObject *obj = state->slice_state.slice_cache;
+    if (obj != NULL) {
+        state->slice_state.slice_cache = NULL;
+        PyObject_GC_Del(obj);
+    }
+}
 
-void _PySlice_Fini(PyInterpreterState *interp)
+void _PySlice_Fini(_PyFreeListState *state)
```
Comment (colesbury) [inline Objects/sliceobject.c:115]: This is identical to `_PySlice_ClearCache`. Do we need both functions? If you want to keep both names, maybe have one call the other.

File: Include/internal/pycore_freelist.h (hunk 1 above)
Comment (colesbury) [reply Include/internal/pycore_freelist.h:75]: It's basically a freelist of max-length 1 🤷

Comment (corona10) [reply Include/internal/pycore_freelist.h:75]: In that case, let's maintain the current naming. (Changing is quite stressful)

File: Objects/sliceobject.c (hunk 2 above)
Comment (corona10) [reply Objects/sliceobject.c:115]: Let's keep both names; I don't want to confuse people who work on the default build.

**Comment by erlend-aasland** [comment]
I'll defer to Sam; please take his comments into account :)

**Review by colesbury — APPROVED** [review]

**Pull request merged**